UnchainedLauncherCore/src/API/ServerBrowser.cs

```diff
@@ -0,0 +1,248 @@
+        public Ports Ports { get; set; } = new(7777, 3075, 7071);
+        // TODO: The selection of the Mod datatype here is potentially
+        // incorrect. Change it to whatever is easier and works with the backend
+        public Mod[] Mods { get; set; } = Array.Empty<Mod>(); 
+    };
+    // TODO: This part of the API should be stabilized
+    // as-is, there are numerous different kinds of "server" objects depending
+    // on where they're coming from/where they're going
+    public record ServerInfo : C2ServerInfo{
+        public String CurrentMap { get; set; } = "";
+        public int PlayerCount { get; set; } = 0;
+        public int MaxPlayers { get; set; } = 0;
+        public ServerInfo() { }
+        public ServerInfo(C2ServerInfo info, A2S_INFO a2s) : base(info) {
+            Update(a2s);
+        }
+        public bool Update(A2S_INFO a2s)
+        {
+            bool wasChanged = (
+                MaxPlayers != a2s.MaxPlayers
+                || PlayerCount != a2s.Players
+                || CurrentMap != a2s.Map);
+            MaxPlayers = a2s.MaxPlayers;
+            PlayerCount = a2s.Players;
+            CurrentMap = a2s.Map;
+            return wasChanged;
+        }
+    }
+
+    public record UniqueServerInfo : ServerInfo {
+        public String UniqueId { get; set; } = "";
+        public double LastHeartbeat { get; set; } = 0;
+    };
+    
+    public record RegisterServerRequest : ServerInfo
+    {
+        public String LocalIpAddress { get; set; }
+        public RegisterServerRequest(ServerInfo info, string localIpAddress) : base(info)
+        {
+            this.LocalIpAddress = localIpAddress;
+        }
+    }
+
+    public record ResponseServer : UniqueServerInfo
+    {
+        public String LocalIpAddress { get; set; } = "";
+        public String IpAddress { get; set; } = "";
+        public ResponseServer() { }
+        public ResponseServer(UniqueServerInfo info, string localIpAddress, string ipAddress) : base(info)
+        {
+            this.LocalIpAddress=localIpAddress;
+            this.IpAddress=ipAddress;
+        }
+    }
+    public record RegisterServerResponse(
+        double RefreshBefore,
+        String Key,
+        ResponseServer Server
+    );
+    public record UpdateServerRequest(
+        int PlayerCount,
+        int MaxPlayers,
+        String CurrentMap
+    );
+
+    public record UpdateServerResponse(double RefreshBefore, ResponseServer Server);
+
+    record GetServersResponse(
+        UniqueServerInfo[] Servers  
+    );
+    //URI is expected to have the /api/v1/ stuff. These functions only append the immediately relevant paths
+    public static class ServerBrowser
+    {
+        static readonly HttpClient httpc = new();
+        private static readonly JsonSerializerOptions sOptions = new(){
+                PropertyNamingPolicy = new SnakeCaseNamingPolicy(),
+                IncludeFields = true
+            };
+        const int TimeOutMillis = 4000;
```

Comment:
This timeout should also be a parameter provided to the ServerBrowser... Though I'm surprised that the HttpClient does not have builtin configuration and management for request timeouts.

Reply:
it does: https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.timeout?view=net-8.0

I think you can use that instead of the cancellation token stuff below.